src/query/service/src/sql/optimizer/heuristic/where_optimizer.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_exception::Result;
+
+use crate::sql::optimizer::ColumnSet;
+use crate::sql::optimizer::SExpr;
+use crate::sql::plans::Filter;
+use crate::sql::plans::LogicalGet;
+use crate::sql::plans::PatternPlan;
+use crate::sql::plans::Prewhere;
+use crate::sql::plans::RelOp;
+use crate::sql::plans::Scalar;
+use crate::sql::MetadataRef;
+
+pub struct WhereOptimizer {
+    metadata: MetadataRef,
+    pattern: SExpr,
+}
+
+impl WhereOptimizer {
+    pub fn new(metadata: MetadataRef) -> Self {
+        Self {
+            metadata,
+            pattern: SExpr::create_unary(
+                PatternPlan {
+                    plan_type: RelOp::Filter,
+                }
+                .into(),
+                SExpr::create_leaf(
+                    PatternPlan {
+                        plan_type: RelOp::LogicalGet,
+                    }
+                    .into(),
+                ),
+            ),
+        }
+    }
+
+    fn collect_columns_no_subqueries(expr: &Scalar, columns: &mut ColumnSet) {
```

Comment:
Maybe `collect_columns` is ok?  We just need some annotations to explain subqueries aren't considered.

Reply:
Ok. This naming is just following ClickHouse.